clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,202 @@
+}
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
+	numSuccess                 int
+	numFailures                int
+	prevProbeFailure           bool
+	maxConsecutiveFailedProbes int
+	consecutiveFailedProbes    int
+}
+
+type apiAvailabilityMeasurement struct {
+	isRunning              bool
+	stopCh                 chan struct{}
+	hosts                  []string
+	apiAvailabilityMetrics *apiAvailabilityMetrics
+	wg                     sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) apiServerAvailable(c clientset.Interface, SSHToMasterSupported bool, config *measurement.MeasurementConfig) bool {
+	for _, host := range a.hosts {
+		// Get the api-server measurement data over SSH.
+		// Start by checking that the provider allows us to do so.
+		if !SSHToMasterSupported {
+			_, err := c.CoreV1().RESTClient().Get().AbsPath("/healthz").DoRaw()
+			if err != nil {
+				return false
+			}
+
+			break
+		}
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		_, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (a *apiAvailabilityMeasurement) start(config *measurement.MeasurementConfig, SSHToMasterSupported bool, probeDuration int) ([]measurement.Summary, error) {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return nil, goerrors.Errorf("Api Server measurement can't execute due to no MasterIps")
+	}
+
+	k8sClient := config.ClusterFramework.GetClientSets().GetClient()
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.wg.Add(1)
+	go func() {
+		defer a.wg.Done()
+		for {
+			select {
+			case <-a.stopCh:
+				return
+			case <-time.After(time.Duration(probeDuration)):
+				apiServerAvailable := a.apiServerAvailable(k8sClient, SSHToMasterSupported, config)
+				// Always update the Probes count
+				a.apiAvailabilityMetrics.numProbes++
+				if apiServerAvailable {
+					updateSuccessMetrics(a.apiAvailabilityMetrics)
+				} else {
+					updateFailureMetrics(a.apiAvailabilityMetrics)
+				}
+				return
+			}
+
+		}
+	}()
+	return nil, nil
+}
+
+func updateFailureMetrics(apiAvailabilityMetrics *apiAvailabilityMetrics) {
+	// increment total number of failed probes count, current number of failed probes by 1
+	apiAvailabilityMetrics.numFailures++
+	apiAvailabilityMetrics.consecutiveFailedProbes++
+	// if the previous probe was not a failure, then set the previous probe failure flag to true.
+	// This would be used by next probe
+	if !apiAvailabilityMetrics.prevProbeFailure {
+		apiAvailabilityMetrics.prevProbeFailure = true
+	}
+}
+
+func updateSuccessMetrics(apiAvailabilityMetrics *apiAvailabilityMetrics) {
+	// if the probe succeeded, then increase the number of successful probes count by 1
+	apiAvailabilityMetrics.numSuccess++
+	// if this successful probe was received after failing probe(s)
+	if apiAvailabilityMetrics.prevProbeFailure {
```

Comment:
This can be simplified a lot - see my comment above.

Reply:
Same as above.